docs/proposals/nodesets.md

```diff
@@ -0,0 +1,157 @@
+directly, nor change the instance type used.  Many people resort to alternative methods of installation,
+to varying degrees of success.
+
+NodeSets put node launching under the control of the master.  They allow the operator to create
+and destroy NodeSets dynamically, or to change the number of instances in a NodeSet.  They
+allow for easier upgrades by launching nodes with different OS/k8s versions.
+
+This simplifies the Kubernetes installation, because the primary requirement is simply to launch a master.  Then
+the installation script can either automatically create a NodeSet, or leave this up to the operator post-creation.
+
+## Use Cases
+
+### Resize number of Nodes through the API
+
+By updating the count on a NodeSet, nodes can be launched or destroyed.  Users will no longer have
+to go to the underlying cloud API.  Kubernetes can likely make better decisions about which instance to
+terminate than a cloud-implemented auto-scaling group can, because it is more aware of the workloads.
+
+### Consistent auto-scaling across providers
+
+In future we can add auto-scaling functionality at the Kubernetes level (instead of at the cloud level), so
+that it works consistently across clouds, and can be based on Kubernetes metrics instead of cloud-metrics
+(i.e. it can take account of pods _not_ scheduled).
+
+### Support multiple NodeSets of different instance types
+
+Users may want to run instances of different types in the same cluster, permanently or transiently.
+Permanently mixing instance types may require more scheduler support, but for transient usage (changing
+instance sizes) this is helpful for changing a cluster.
+
+Users may also want to mix instance types in a way that is not visible to the cluster, for example mixing
+on-demand and spot instances on AWS.
+
+### Support upgrading by specifying software versions
+
+By specifying that a new NodeSet should be launched with a different image or kubernetes version,
+users can easily perform an online upgrade of their nodes (within compatible versions).
+
+## Functionality
+
+The NodeSet supports standard API features:
+- create
+  - The spec for NodeSet includes a count of nodes
+  - The spec includes a template, which specifies the type of nodes that will be launched as labels.
+    We imagine that a cloudprovider actually has a pool of machines of various instance types, images,
+    kubernetes versions etc.  We will label each node with the labels describing its instance type, running image,
+    kubernetes version etc on creation.  Then the selector acts to specify the information that nodes must have.
+    Thus, specifying `kubernetes.io/instanceType=n1-highmem-16` will ensure that nodes are launched with
+    `n1-highmem-16` instance type.  Selectors with the `kubernetes.io/` prefix must be known, or it is a
+    validation error; selectors without the prefix are treated as normal labels.
+    (TODO: Is this too 'cute'?)
```

Comment:
I would not want to use labels in this way. 

I agree that baked-in types, as with volume sources, is not the way we want to go.

I'd rather figure out a first-class extension mechanism, but annotations are the default answer.

#14095 is also worth a look if you haven't seen it.